lib/Cro/Iri.pm6

```diff
@@ -1,75 +1,253 @@
-use Cro::Uri;
+use Cro::Uri :decode-percents, :encode-percents;
+use Cro::ResourceIdentifier;
 
-class Cro::Iri is Cro::Uri {
+class X::Cro::Iri::ParseError is Exception {
+    has $.reason = 'malformed syntax';
+    has $.iri-string is required;
+    method message() {
+        "Unable to parse IRI '$!iri-string': $!reason"
+    }
+}
+
+class Cro::Iri does Cro::ResourceIdentifier {
     grammar GenericParser is Cro::Uri::GenericParser {
```

Comment:
Do we actually use anything that is inherited here? Same question for action class.

Reply:
We reuse `scheme`, `port`, `IPv4address`, and `IPv6address` (and several other rules those use).